### PR TITLE
AWS Batch Compute Environment add `update_policy` parameter

### PR DESCRIPTION
**Pull request by drewmullen**
Closes: #32927, #34145

## Resource

```shell
$ make testacc TESTS=TestAccBatchComputeEnvironment_ PKG=batch
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/batch/... -v -count 1 -parallel 20 -run='TestAccBatchComputeEnvironment_'  -timeout 360m
=== RUN   TestAccBatchComputeEnvironment_basic
=== PAUSE TestAccBatchComputeEnvironment_basic
=== RUN   TestAccBatchComputeEnvironment_disappears
=== PAUSE TestAccBatchComputeEnvironment_disappears
=== RUN   TestAccBatchComputeEnvironment_nameGenerated
=== PAUSE TestAccBatchComputeEnvironment_nameGenerated
=== RUN   TestAccBatchComputeEnvironment_namePrefix
=== PAUSE TestAccBatchComputeEnvironment_namePrefix
=== RUN   TestAccBatchComputeEnvironment_eksConfiguration
=== PAUSE TestAccBatchComputeEnvironment_eksConfiguration
=== RUN   TestAccBatchComputeEnvironment_createEC2
=== PAUSE TestAccBatchComputeEnvironment_createEC2
=== RUN   TestAccBatchComputeEnvironment_updatePolicyCreate
=== PAUSE TestAccBatchComputeEnvironment_updatePolicyCreate
=== RUN   TestAccBatchComputeEnvironment_updatePolicyUpdate
=== PAUSE TestAccBatchComputeEnvironment_updatePolicyUpdate
=== RUN   TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
=== PAUSE TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
=== RUN   TestAccBatchComputeEnvironment_createSpot
=== PAUSE TestAccBatchComputeEnvironment_createSpot
=== RUN   TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage
=== PAUSE TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage
=== RUN   TestAccBatchComputeEnvironment_createFargate
=== PAUSE TestAccBatchComputeEnvironment_createFargate
=== RUN   TestAccBatchComputeEnvironment_createFargateSpot
=== PAUSE TestAccBatchComputeEnvironment_createFargateSpot
=== RUN   TestAccBatchComputeEnvironment_updateState
=== PAUSE TestAccBatchComputeEnvironment_updateState
=== RUN   TestAccBatchComputeEnvironment_updateServiceRole
=== PAUSE TestAccBatchComputeEnvironment_updateServiceRole
=== RUN   TestAccBatchComputeEnvironment_defaultServiceRole
=== PAUSE TestAccBatchComputeEnvironment_defaultServiceRole
=== RUN   TestAccBatchComputeEnvironment_ComputeResources_minVCPUs
=== PAUSE TestAccBatchComputeEnvironment_ComputeResources_minVCPUs
=== RUN   TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs
=== PAUSE TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs
=== RUN   TestAccBatchComputeEnvironment_ec2Configuration
=== PAUSE TestAccBatchComputeEnvironment_ec2Configuration
=== RUN   TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup
=== PAUSE TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup
=== RUN   TestAccBatchComputeEnvironment_launchTemplate
=== PAUSE TestAccBatchComputeEnvironment_launchTemplate
=== RUN   TestAccBatchComputeEnvironment_updateLaunchTemplate
=== PAUSE TestAccBatchComputeEnvironment_updateLaunchTemplate
=== RUN   TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate
=== PAUSE TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate
=== RUN   TestAccBatchComputeEnvironment_tags
=== PAUSE TestAccBatchComputeEnvironment_tags
=== RUN   TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources
=== PAUSE TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources
=== RUN   TestAccBatchComputeEnvironment_updateEC2
=== PAUSE TestAccBatchComputeEnvironment_updateEC2
=== RUN   TestAccBatchComputeEnvironment_createEC2WithoutComputeResources
=== PAUSE TestAccBatchComputeEnvironment_createEC2WithoutComputeResources
=== CONT  TestAccBatchComputeEnvironment_basic
=== CONT  TestAccBatchComputeEnvironment_updateServiceRole
=== CONT  TestAccBatchComputeEnvironment_createEC2WithoutComputeResources
=== CONT  TestAccBatchComputeEnvironment_createEC2
=== CONT  TestAccBatchComputeEnvironment_ec2Configuration
=== CONT  TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources
=== CONT  TestAccBatchComputeEnvironment_ComputeResources_minVCPUs
=== CONT  TestAccBatchComputeEnvironment_tags
=== CONT  TestAccBatchComputeEnvironment_eksConfiguration
=== CONT  TestAccBatchComputeEnvironment_updatePolicyUpdate
=== CONT  TestAccBatchComputeEnvironment_updateEC2
=== CONT  TestAccBatchComputeEnvironment_createFargateSpot
=== CONT  TestAccBatchComputeEnvironment_createFargate
=== CONT  TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage
=== CONT  TestAccBatchComputeEnvironment_createSpot
=== CONT  TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
=== CONT  TestAccBatchComputeEnvironment_updateLaunchTemplate
=== CONT  TestAccBatchComputeEnvironment_updatePolicyCreate
=== CONT  TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup
=== CONT  TestAccBatchComputeEnvironment_updateState
--- PASS: TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources (144.33s)
=== CONT  TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate
--- PASS: TestAccBatchComputeEnvironment_createEC2WithoutComputeResources (637.26s)
=== CONT  TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs
--- PASS: TestAccBatchComputeEnvironment_basic (1007.97s)
=== CONT  TestAccBatchComputeEnvironment_defaultServiceRole
--- PASS: TestAccBatchComputeEnvironment_createFargate (1131.06s)
=== CONT  TestAccBatchComputeEnvironment_nameGenerated
--- PASS: TestAccBatchComputeEnvironment_createFargateSpot (1164.89s)
=== CONT  TestAccBatchComputeEnvironment_disappears
--- PASS: TestAccBatchComputeEnvironment_createEC2 (1211.40s)
=== CONT  TestAccBatchComputeEnvironment_namePrefix
--- PASS: TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage (1281.13s)
=== CONT  TestAccBatchComputeEnvironment_launchTemplate
--- PASS: TestAccBatchComputeEnvironment_ec2Configuration (1400.07s)
--- PASS: TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup (1498.77s)
--- PASS: TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags (1517.84s)
--- PASS: TestAccBatchComputeEnvironment_updatePolicyUpdate (1649.08s)
--- PASS: TestAccBatchComputeEnvironment_updateState (1661.37s)
--- PASS: TestAccBatchComputeEnvironment_updatePolicyCreate (1661.56s)
--- PASS: TestAccBatchComputeEnvironment_createSpot (1702.94s)
--- PASS: TestAccBatchComputeEnvironment_defaultServiceRole (843.08s)
--- PASS: TestAccBatchComputeEnvironment_updateServiceRole (1889.38s)
--- PASS: TestAccBatchComputeEnvironment_nameGenerated (787.80s)
--- PASS: TestAccBatchComputeEnvironment_disappears (784.71s)
--- PASS: TestAccBatchComputeEnvironment_namePrefix (744.84s)
--- PASS: TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate (1848.79s)
--- PASS: TestAccBatchComputeEnvironment_updateLaunchTemplate (2014.04s)
--- PASS: TestAccBatchComputeEnvironment_tags (2019.20s)
--- PASS: TestAccBatchComputeEnvironment_launchTemplate (795.59s)
--- PASS: TestAccBatchComputeEnvironment_eksConfiguration (2218.33s)
--- PASS: TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs (1623.90s)
--- PASS: TestAccBatchComputeEnvironment_ComputeResources_minVCPUs (2263.76s)
--- PASS: TestAccBatchComputeEnvironment_updateEC2 (2337.19s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/batch      2339.211s
```

## Data Source

```shell
make testacc TESTS=TestAccBatchComputeEnvironmentDataSource_ PKG=batch
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/batch/... -v -count 1 -parallel 20 -run='TestAccBatchComputeEnvironmentDataSource_'  -timeout 360m
=== RUN   TestAccBatchComputeEnvironmentDataSource_basic
=== PAUSE TestAccBatchComputeEnvironmentDataSource_basic
=== RUN   TestAccBatchComputeEnvironmentDataSource_basicUpdatePolicy
=== PAUSE TestAccBatchComputeEnvironmentDataSource_basicUpdatePolicy
=== CONT  TestAccBatchComputeEnvironmentDataSource_basic
=== CONT  TestAccBatchComputeEnvironmentDataSource_basicUpdatePolicy
--- PASS: TestAccBatchComputeEnvironmentDataSource_basic (156.73s)
--- PASS: TestAccBatchComputeEnvironmentDataSource_basicUpdatePolicy (192.10s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/batch      194.059s
```

### CODE REVIEW

**Review by AdamTylerLynch — APPROVED** [review]
LGTM 🚀

File: internal/service/batch/compute_environment.go [hunk 1]
```diff
@@ -300,6 +317,23 @@ func resourceComputeEnvironmentCreate(ctx context.Context, d *schema.ResourceDat
...
 
+	// UpdatePolicy is not possible to set with CreateComputeEnvironment
+	if v, ok := d.GetOk("update_policy"); ok && len(v.([]interface{})) > 0 && v.([]interface{})[0] != nil {
+		inputeUpdateOnCreate := &batch.UpdateComputeEnvironmentInput{
+			ComputeEnvironment: aws.String(d.Id()),
+			UpdatePolicy:       expandComputeEnvironmentUpdatePolicy(v.([]interface{})),
+		}
+		log.Printf("[DEBUG] Creating Batch Compute Environment extra arguments: %s", input)
+
+		if _, err := conn.UpdateComputeEnvironmentWithContext(ctx, inputeUpdateOnCreate); err != nil {
```
Comment (jar-b) [inline internal/service/batch/compute_environment.go:328]:
```suggestion
		inputUpdateOnCreate := &batch.UpdateComputeEnvironmentInput{
			ComputeEnvironment: aws.String(d.Id()),
			UpdatePolicy:       expandComputeEnvironmentUpdatePolicy(v.([]interface{})),
		}
		log.Printf("[DEBUG] Creating Batch Compute Environment extra arguments: %s", inputUpdateOnCreate)

		if _, err := conn.UpdateComputeEnvironmentWithContext(ctx, inputUpdateOnCreate); err != nil {
```

File: website/docs/d/batch_compute_environment.html.markdown [hunk 2]
```diff
@@ -36,4 +36,5 @@ This data source exports the following attributes in addition to the arguments a
 * `status` - Current status of the compute environment (for example, `CREATING` or `VALID`).
 * `status_reason` - Short, human-readable string to provide additional details about the current status of the compute environment.
 * `state` - State of the compute environment (for example, `ENABLED` or `DISABLED`). If the state is `ENABLED`, then the compute environment accepts jobs from a queue and can scale out automatically based on queues.
+* `update_policy` - (Optional) Specifies the infrastructure update policy for the compute environment.
```
Comment (jar-b) [inline website/docs/d/batch_compute_environment.html.markdown:39]:
```suggestion
* `update_policy` - Specifies the infrastructure update policy for the compute environment.
```

**Review by jar-b — COMMENTED** [review]
Looks good! I wasn't able to commit to the downstream branch, so left a couple suggestions with what I'd changed locally. Just one small typo/logging fix and a documentation fix. 

Will run the full test suite and merge after the changes go through.

**Comment by drewmullen** [comment]
@jar-b how did you do multi-line `suggestion boxes?

image

**Comment by drewmullen** [comment]
Also, going foward we wont use orgs for PR sources and instead will use personal repos. That is whats preventing you from committing back upstream - apologies for that

**Comment by jar-b** [comment]
> @jar-b how did you do multi-line `suggestion boxes?

Just to the left of the -/+ diff markers there is a blue box you can drag across multiple lines.

image

**Comment by jar-b** [comment]
No problem

> Also, going foward we wont use orgs for PR sources and instead will use personal repos. That is whats preventing you from committing back upstream - apologies for that

No problem - thanks for getting the changes in. Running tests now.

**Review by jar-b — APPROVED** [review]
LGTM 🎉 

```console
% make testacc PKG=batch TESTS="TestAccBatchComputeEnvironment_|TestAccBatchComputeEnvironmentDataSource_"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/batch/... -v -count 1 -parallel 20 -run='TestAccBatchComputeEnvironment_|TestAccBatchComputeEnvironmentDataSource_'  -timeout 360m

--- PASS: TestAccBatchComputeEnvironment_createUnmanagedWithComputeResources (4.51s)
=== CONT  TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate
--- PASS: TestAccBatchComputeEnvironment_createEC2WithoutComputeResources (32.74s)
=== CONT  TestAccBatchComputeEnvironment_ec2Configuration
--- PASS: TestAccBatchComputeEnvironment_createFargate (37.77s)
=== CONT  TestAccBatchComputeEnvironment_updateLaunchTemplate
--- PASS: TestAccBatchComputeEnvironmentDataSource_basic (40.38s)
=== CONT  TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs
--- PASS: TestAccBatchComputeEnvironment_CreateSpotAllocationStrategy_bidPercentage (52.92s)
=== CONT  TestAccBatchComputeEnvironment_launchTemplate
=== CONT  TestAccBatchComputeEnvironment_updatePolicyUpdate
--- PASS: TestAccBatchComputeEnvironment_disappears (52.99s)
--- PASS: TestAccBatchComputeEnvironmentDataSource_basicUpdatePolicy (54.44s)
=== CONT  TestAccBatchComputeEnvironment_defaultServiceRole
=== NAME  TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
    compute_environment_test.go:481: Step 1/2 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout


        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place

        Terraform will perform the following actions:

          # aws_batch_compute_environment.test will be updated in-place
          ~ resource "aws_batch_compute_environment" "test" {
                id                       = "tf-acc-test-8202777925839378489"
                tags                     = {}
                # (9 unchanged attributes hidden)

              ~ compute_resources {
                  ~ desired_vcpus      = 4 -> 8
                    tags               = {
                        "key1" = "value1"
                    }
                    # (10 unchanged attributes hidden)

                    # (1 unchanged block hidden)
                }
            }

        Plan: 0 to add, 1 to change, 0 to destroy.
--- PASS: TestAccBatchComputeEnvironment_createEC2 (64.28s)
=== CONT  TestAccBatchComputeEnvironment_updateServiceRole
--- PASS: TestAccBatchComputeEnvironment_nameGenerated (65.61s)
=== CONT  TestAccBatchComputeEnvironment_basic
--- PASS: TestAccBatchComputeEnvironment_namePrefix (66.60s)
--- PASS: TestAccBatchComputeEnvironment_createFargateSpot (67.08s)
--- PASS: TestAccBatchComputeEnvironment_ec2ConfigurationPlacementGroup (68.61s)
--- PASS: TestAccBatchComputeEnvironment_updateState (70.35s)
--- PASS: TestAccBatchComputeEnvironment_ec2Configuration (38.36s)
--- PASS: TestAccBatchComputeEnvironment_UpdateSecurityGroupsAndSubnets_fargate (73.23s)
--- PASS: TestAccBatchComputeEnvironment_tags (78.15s)
--- PASS: TestAccBatchComputeEnvironment_updatePolicyCreate (85.95s)
--- PASS: TestAccBatchComputeEnvironment_defaultServiceRole (32.77s)
--- PASS: TestAccBatchComputeEnvironment_launchTemplate (35.03s)
--- PASS: TestAccBatchComputeEnvironment_basic (29.76s)
--- PASS: TestAccBatchComputeEnvironment_updateLaunchTemplate (59.30s)
--- PASS: TestAccBatchComputeEnvironment_updatePolicyUpdate (45.57s)
--- PASS: TestAccBatchComputeEnvironment_updateEC2 (107.14s)
--- PASS: TestAccBatchComputeEnvironment_ComputeResources_maxVCPUs (69.31s)
--- PASS: TestAccBatchComputeEnvironment_updateServiceRole (64.83s)
--- PASS: TestAccBatchComputeEnvironment_ComputeResources_minVCPUs (201.94s)
--- FAIL: TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags (284.80s)
--- PASS: TestAccBatchComputeEnvironment_createSpot (472.97s)
--- PASS: TestAccBatchComputeEnvironment_eksConfiguration (713.40s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-aws/internal/service/batch      716.838s
```

Failed test is unrelated to these changes and transient, passing after re-running:

```console
% make testacc PKG=batch TESTS=TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/batch/... -v -count 1 -parallel 20 -run='TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags'  -timeout 360m
=== RUN   TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
=== PAUSE TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
=== CONT  TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags
--- PASS: TestAccBatchComputeEnvironment_CreateEC2DesiredVCPUsEC2KeyPairImageID_computeResourcesTags (65.30s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/batch      68.466s
```

**Comment by jar-b** [comment]
Thanks for your contribution, @drewmullen! 👍

**Pull request merged**